lib.packagesFromDirectoryRecursive: use explicit recursion, support nested scopes

nicoonoclaste opened the pull request:
Follow-up on #359941

- ~~Extend `lib.filesystem.packagesFromDirectoryRecursive` with support for explicit recursion.~~
- ~~Add an example showing how to use `recurseIntoDirectory`.~~
- [x] Update `packagesFromDirectoryRecursive` to create nested scopes for each (sub)directory (not just the top-level one) when `newScope` is given.
- ~~Apply `recurseIntoAttrs` by default when recursing into a directory.~~ (Deferred to follow-up PR)
- [x] Add a test using scopes for `packagesFromDirectoryRecursive`, ensuring:
  - packages can depend on other packages in the same scope *and* in ancestor scopes (the scope's parent or grandparent) ;
  - packages can have mutual dependencies.
- [x] Reject unknown parameters (breaking change!)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested `nix-instantiate --eval --strict lib/tests/misc.nix`
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] [25.05 Release Note](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nicoonoclaste committed:
lib/tests: move `packages-from-directory/*` into a subdir

in preparation for adding more tests for `lib.packagesFromDirectoryRecursive`

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: use explicit recursion, support nested scopes

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: add tests for nested scopes

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: default to creating new scopes when `newScope` is provided

 Co-authored-by: Rebecca Turner <rbt@sent.as>

9999years commented on lib/filesystem.nix:
```diff
@@ -361,32 +360,49 @@ in
+      recurseIntoDirectory ?
+        if args ? newScope then
+          # when `newScope` is provided, create a new scope when recursing into a directory (incl. the top one)
+          f: { newScope, ... }@args:
+            lib.recurseIntoAttrs (lib.makeScope newScope (self:
+              f (args // {
+                inherit (self) callPackage newScope;
+              })
+            ))
```
This sort of pattern is really opaque and challenging to read. I'd like to have some more comments explaining what's going on here. (E.g., "A scope means that packages in the same directory can refer to each other...").

Also, I'd really prefer if `f` had a more descriptive name. Any of `inner`, `helper`, or `recurse` would be fine.

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: hoist `processDir` out of the lambda

Fixes a bug preventing `recurseIntoDirectory` from changing the `directory` argument.

Moreover, `processDir` now cannot capture arguments from the `packagesFromDirectoryRecursive` call,
entirely preventing this class of bug from reoccurring should new parameters be added etc.

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: reject unknown arguments

see [discussion](https://github.com/NixOS/nixpkgs/pull/270537#discussion_r1862458435)

9999years commented on lib/filesystem.nix:
```diff
@@ -306,106 +309,154 @@ in
+    let
+      inherit (lib.path) append;
+
+      # generate an attrset corresponding to a given directory
+      # this function is outside `packagesFromDirectoryRecursive`'s lambda expression,
+      #  to prevent accidentally using its parameters
```
```suggestion
      # Generate an attrset corresponding to a given directory.
      # This function is outside `packagesFromDirectoryRecursive`'s lambda expression
      # to prevent accidentally using its parameters.
```

Non-blocking slight reformatting. (I got this style preference from https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html)

9999years commented on lib/filesystem.nix:
```diff
@@ -306,106 +309,154 @@ in
+          # `processDir` is the same function as defined above
+          # `args` are the arguments passed to (this recursive call of) `packagesFromDirectoryRecursive`
+          processDir: { newScope, ... }@args:
+            # create a new scope and mark it `recurseForDerivations`; see
```
```suggestion
            # Create a new scope and mark it `recurseForDerivations`.
            # This lets the packages refer to each other.
            # See:
```

9999years reviewed (approved):
LGTM. Left a couple style comments but nothing substantial

nicoonoclaste committed:
lib.filesystem: remove dead `let`-bindings

Found by `deadnix` 🖤

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: Add example use of `recurseIntoDirectory`

hsjobeki commented on lib/filesystem.nix:
```diff
@@ -372,46 +393,100 @@ in
+      # recurseIntoDirectory can modify the function used when processing directory entries
+      #  and recurseArgs can (optionally) hold data for its use ; see nixdoc above
+      recurseArgs ? throw "lib.packagesFromDirectoryRecursive: recurseArgs wasn't passed in args",
+      recurseIntoDirectory ?
```
This function could be its own unit. For example `defaultRecurseIntoDirectory`. This makes it possible to unit-test and reuse this function. I think inline here is a little bit unhappy.

Also this inline documentation as plain comment will not render into any documentation.

If you pull this out you can write a nice doc-comment describing the behavior.
Place mutual references in the parent `function-library-lib.filesets.packagesFromDirectoryRecursive`
This will make it much easier for people to understand and implement their own function if needed.
```nix
/** 
  Create a new scope and mark it `recurseForDerivations`.
  This lets the packages refer to each other.
  
  See also:
  - [lib.makeScope](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.customisation.makeScope) and
  - [lib.recurseIntoAttrs](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.customisation.makeScope)
*/
defaultRecurseIntoDirectory = args: if args ? newScope then ...
```

hsjobeki replied:
How about pulling this into the let block above? So it is more clear what variables are in the scope of this function.

> the default is a poor example, as it mixes two distinct usecases (with and without scopes) so I'd be more in favour of adding an example showing how to implement scopes manually.

I think you have one example in the docs above?

hsjobeki commented on lib/filesystem.nix:
```diff
@@ -372,46 +393,100 @@ in
-      ...
-    }:
+      # recurseIntoDirectory can modify the function used when processing directory entries
+      #  and recurseArgs can (optionally) hold data for its use ; see nixdoc above
```
```suggestion
      #  and recurseArgs can (optionally) hold data for its use ; see comment above
```

hsjobeki replied:
Yes please. The term `nixdoc` refers to https://github.com/nix-community/nixdoc.

hsjobeki commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -310,3 +310,11 @@
+  [`lib.packagesFromDirectoryRecursive`]: https://nixos.org/manual/nixpkgs/stable/#function-library-lib.filesystem.packagesFromDirectoryRecursive
+  [`lib.recurseIntoAttrs`]: https://nixos.org/manual/nixpkgs/stable/#function-library-lib.attrsets.recurseIntoAttrs
+
+### Other notable changes {#sec-release-25.05-lib-notable-changes}
```
It seems we used a different heading in all the previous release notes:

```suggestion
### Additions and Improvements {#sec-release-25.05-lib-additions-improvements}
```

hsjobeki reviewed (dismissed):
Nice changes. Thank you for the efforts that you already put into it.

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: use explicit recursion, support nested scopes

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: default to creating new scopes when `newScope` is provided

hsjobeki commented on lib/filesystem.nix:
```diff
@@ -372,46 +393,100 @@ in
+            "${removeSuffix ".nix" name}" = callPackage path {};
+          } else if type == "regular" then {
+            # ignore non-nix files
+          } else throw ''
```
Could you document that only `regular` files are supported? Symlinks and other file types present in a directory would trigger this error I assume.

https://hydra.nixos.org/build/278148859/download/1/manual/language/builtins.html#builtins-readDir

hsjobeki commented on lib/filesystem.nix:
```diff
@@ -372,46 +393,100 @@ in
+      recurseIntoDirectory = processDir: args: let
+        result = processDir args;
+      in result // {
+        recurseForDerivations = with lib;
+          any (child: isDerivation child || child.recurseForDerivations or false) result;
```
```suggestion
        recurseForDerivations = lib.any (child: lib.isDerivation child || child.recurseForDerivations or false) result;
```

With statements are bad we shouldn't encurage their usage as people tend to copy examples.

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: document that symlinks are ignored

This was implicit in the previous “other files are ignored.”

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: Add example use of `recurseIntoDirectory`

nicoonoclaste committed:
squash! lib.packagesFromDirectoryRecursive: hoist `processDir` out of the lambda

also hoisted the default `recurseIntoDirectory`, as requested during review

infinisil reviewed (changes requested):
Thanks for the PR! Me and @hsjobeki reviewed this PR more closely together in a call and agreed on this feedback:
- The new `newScope` attribute argument and the removal of `...` from `packagesFromDirectoryRecursive` looks good to us.
- However, the `recurseIntoDirectory` concept feels a bit messy and hard to understand (we didn't quite get it in our ~1 hour of reviewing this PR together). In particular:
  - It's unclear what `processDir` does based on the documentation
  - `recurseArgs` has unclear motivation, is not tested and is missing example documentation.

In alignment with the [one PR for each change](https://github.com/NixOS/nixpkgs/tree/master/lib#make-one-pr-for-each-change) `lib` PR guideline, we think this PR should defer the `recurseIntoDirectory` and `recurseArgs` introduction to a separate PR that clearly states motivation for the new interfaces and adds missing docs with examples.

Also note that there's now [separate Nixpkgs release notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2505.section.md#nixpkgs-library-sec-nixpkgs-release-2505-lib) where you need to note the `lib` changes, which is the cause of the merge conflict.

nicoonoclaste committed:
lib.packagesFromDirectoryRecursive: drop open recursion from API surface

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: drop open recursion from API surface

infinisil commented:
> I'm somewhat confused there, since adding open recursion is part of how nested scopes are implemented (at least in this PR) but I can remove it from the public API and simplify as much as possible.

The `makeScope` is already a useful change on its own. The fact that it's more limited allows guaranteeing more about the result in a more abstract sense. I like the saying "Constraints Liberate, Liberties Constrain" here. Of course, it's always possible to remove constraints later (by adding `recurseIntoDirectory`), but we can't go the other way and add more constraints (that would be a breaking change). So I think it's a good idea to start out with more constraints and loosen them as necessary :)

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: drop open recursion from API surface

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: drop open recursion from API surface

infinisil commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -310,3 +310,11 @@
   - [`lib.types.enum`](https://nixos.org/manual/nixos/unstable/#sec-option-types-basic): Previously the `functor.payload` was the list of enum values directly. Now it is an attribute set containing the values in the `values` attribute.
   - [`lib.types.separatedString`](https://nixos.org/manual/nixos/unstable/#sec-option-types-string): Previously the `functor.payload` was the seperator directly. Now it is an attribute set containing the seperator in the `sep` attribute.
+
+- [`lib.packagesFromDirectoryRecursive`] now rejects unknown arguments, and applies [`lib.recurseIntoAttrs`] when recursing into a directory.
```
The current code only applies `recurseIntoAttrs` if `makeScope` is passed. That seems like an oversight.

infinisil commented on lib/filesystem.nix:
```diff
@@ -372,46 +369,59 @@ in
+      # See:
+      #  [lib.makeScope](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.customisation.makeScope) and
+      #  [lib.recurseIntoAttrs](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.customisation.makeScope)
+      recurseIntoAttrs (makeScope newScope (self:
+        # generate the attrset representing the directory, using the new scope's `callPackage` and `newScope`
+        processDir (args // {
+          inherit (self) callPackage newScope;
+        })
+      ))
+    else
+      processDir args
```
Probably makes sense to move `recurseIntoAttrs` into `processDir` to apply it unconditionally.

nicoonoclaste committed:
fixup! lib.packagesFromDirectoryRecursive: reject unknown arguments

hsjobeki commented:
Generally looking good to me :+1: 
Could you rebase your changes. We might merge this one tomorrow. :)

hsjobeki reviewed (approved):
Approving based on the changes. Still needs to be rebased to fix the conflicts and get rid of the fixup commits.
Well done :tada:

infinisil reviewed (approved):
Same as @hsjobeki above!

katexochen commented:
@nbraud can you squash in the fixups and resolve the merge confilct? Looks like this is good to merge otherwise. :)

katexochen commented on lib/filesystem.nix:
```diff
@@ -372,46 +369,59 @@ in
+    in
     {
       callPackage,
+      newScope ? throw "lib.packagesFromDirectoryRecursive: newScope wasn't passed in args",
```
Can we default to the newScope from callPackage?

```suggestion
      newScope ? (callPackage ({ newScope }: { inherit newScope; }) { }).newScope,
```

hsjobeki replied:
@katexochen Do you want to take that over. It seems the PR maintainer is unresponsive. Can you squash the original commits into a new pr and link this one so we can move on?

hsjobeki replied:
Regarding your suggestion. I'd prefer to merge what is already reviewed. Its easy to make a follow up pr for that.

katexochen replied:
Let's give them a day or two to react. Otherwise I'll take over.

katexochen replied:
Yes, I'll take over, but it's a bit fiddly, the fixups don't squash cleanly and I have to resolve merge conflicts with master, but I'm on it.

katexochen replied:
@nbraud looking at the code now, I think my suggestion doesn't make any sense as this throw is never triggered? The way I understand it, if now newScope is passed we will not use nested scopes, so there is no point of providing a default for newScope.

llakala commented:
Merged in #392800
